### PR TITLE
Add `hook/` to strict dependencies check

### PR DESCRIPTION
**Pull request by dcharkes**
* https://github.com/dart-lang/native/issues/160

### CODE REVIEW

File: lib/src/validator/strict_dependencies.dart [hunk 1]
```diff
@@ -69,13 +69,13 @@ class StrictDependenciesValidator extends Validator {
...
-  /// Validates that no Dart files in `lib/` or `bin/` have dependencies that
-  /// aren't in [deps].
+  /// Validates that no Dart files in `lib/`, `bin/`, or `hook/` have
+  /// dependencies that aren't in [deps].
   ///
   /// The [devDeps] are used to generate special warnings for files that import
   /// dev dependencies.
   void _validateLibBin(Set<String> deps, Set<String> devDeps) {
-    for (var usage in _usagesBeneath(['lib', 'bin'])) {
+    for (var usage in _usagesBeneath(['lib', 'bin', 'hook'])) {
```
[the remark this answers is not in the document]
Comment (jonasfj) [reply lib/src/validator/strict_dependencies.dart:78]: Just a thought, I'm not actually sure. Perhaps this is fine for now.

@sigurdm thoughts?

[the remark this answers is not in the document]
Comment (dcharkes) [reply lib/src/validator/strict_dependencies.dart:78]: I think this holds for all hooks in the future. The nature of hooks is that we run them with a specific CLI as process invocations from the embedder. Those process calls use the package_config of the root package. We want to ensure that we don't have a different version of the package when a dependency runs a package's hook.

[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply lib/src/validator/strict_dependencies.dart:78]: We might have dev-only hooks, for code generation for example?

**Comment by dcharkes** [comment]
Now with tests!

File: lib/src/validator.dart [hunk 2]
```diff
@@ -224,7 +224,9 @@ abstract class Validator {
     return files
         .where(
           recursive
-              ? (file) => p.isWithin(base, p.canonicalize(file))
+              ? (file) =>
```
Comment (HosseinYousefi) [inline lib/src/validator.dart:227]: The docs and the argument name of this method assumes `dir` to be a directory.

File: lib/src/validator/strict_dependencies.dart [hunk 3]
```diff
@@ -69,13 +69,18 @@ class StrictDependenciesValidator extends Validator {
     _validateBenchmarkTestTool(dependencies, devDependencies);
   }
 
-  /// Validates that no Dart files in `lib/` or `bin/` have dependencies that
-  /// aren't in [deps].
+  /// Validates that no Dart files in `lib/`, `bin/`, or `hook/` have
```
Comment (HosseinYousefi) [inline lib/src/validator/strict_dependencies.dart:72]: Specify the exact files within the hook directory.

**Comment by HosseinYousefi** [comment]
LGTM!

**Review by sigurdm — APPROVED** [review]

**Pull request merged**